### ISSUE TITLE
Error-level flag for newing up an interface

### ISSUE THREAD

**Issue by retailcoder** [rubberduck-vba/Rubberduck#5132]
**What**
A code quality inspection flagging any `IdentifierReference` referring to a class module that Rubberduck understands as an abstract interface.

**Why**
VB6/VBA will not issue any compile-time error if an interface is instantiated. At run-time, this can lead to [hard-to-diagnose, subtle bugs](https://stackoverflow.com/q/57778199/1188513).

**Example**
This code should trigger the inspection:

```vb
'IClass1
'@Interface
Public Sub DoSomething()
End Sub

'Module1
Public Sub DoItWrong()
    Dim foo As IClass1
    Set foo = New IClass1 '<~ here
End Sub
```

---

**QuickFixes**

1. **Instantiate {concrete implementation}**

Each class that implements the newed-up interface should be listed as a possible quick-fix.

    Example code, after quickfix is applied:

    ```vb
'Module1
Public Sub DoItWrong()
    Dim foo As IClass1
    Set foo = New Class1 '<~ where Class1 implements IClass1
End Sub
    ```

---

**Resources**
TBD

**Comment by mansellan** [comment]
It's been a while since I wrote VB Classic in anger, so this may be entirely wrong, but... aren't there legitimate use-cases for doing this. Consider the below which (I think?) *approximates* class inheritance:

```vb
' Base.cls
Public Sub DoSomething()
   MsgBox "Greetings from the default DoSomething()"
End Sub

' Ordinary.cls
Implements Base

Private Base _base

Public Sub Class_Initialize()
   _base = New Base
End Sub

Public Sub Base_DoSomething()
  _base.DoSomething() ' <-- 'Inherit' the base implementation
End Sub

' Extraordinary.cls
Implements Base

Public Sub Base_DoSomething
   MsgBox "This is sort of like an override"
End Sub

**Comment by retailcoder** [comment]
@Mansellan that's basically a malformed decorator pattern - like [here](https://stackoverflow.com/a/57641879/1188513). Why not achieve the same result by implementing an *actual* decorator instead?

**Comment by mansellan** [comment]
Well yes, that's true... But decorator was described from the assumption that interfaces *could not* have implementations. True of Java and C#, not true of VBA. So is it a malformed decorator, or is is a VBC-specific pattern? What are the reasons not to do this, other than confusing people more used to other languages? Genuibne question...

**Comment by mansellan** [comment]
Decorator in VBC requires factory methods, which could be overkill.

**Comment by retailcoder** [comment]
> Decorator in VBC requires factory methods, which could be overkill.

I disagree. Factory methods off the default instance are pretty lightweight, I use them on pretty much every class I write - along with the corresponding explicit interface.

OOP principles and patterns are essentially language-agnostic: to the extent that class inheritance isn't a requirement (e.g. templated methods), I see no reason to VB-ize  patterns that are widely recognized by OOP devs regardless of what language they prefer. VBC doesn't support inheritance - let's not pretend it does, and embrace what OOP has in store that *can* be leveraged. If it looks like a bastardized decorator, then OOP devs will see a bastardized decorator... and since design patterns are a communication device, twisting a pattern to solve another problem than the one it was made to solve,  is inherently a communication failure.

It doesn't take much to take that code and turn it into an actual recognizable decorator. If you're already trying to simulate inheritance, then you're writing something that wants to be OOP; your project likely already has more class modules than the VBE comfortably handles, and thus you're probably already leveraging `@Folder` annotations to regroup modules by functionality - why not use a couple more RD annotations then, and leverage `@PredeclaredId` and `@Interface` as well?

**Comment by retailcoder** [comment]
> What are the reasons not to do this, other than confusing people more used to other languages?

Not warning about concrete implementations in interface modules can (and has - see linked SO thread in the OP) cause bugs/problems that are hard to diagnose.

**Comment by miannizzotto** [comment]
Hi @retailcoder,

Moved by curiosity (I'm the guy who posted this question on SO yesterday) I've had a look at RubberDuck's code for the first time today and I admit I am a very new user to RubberDuck (I unfortunately don't code much in VBA anymore). Congrats for this amazing add-in! 
 
So, I've cloned the RubberDuck repository and enjoying some debugging to understand how the code works.
I don't engage yet as assignee (too soon for someone just jumping into such a huge project :), but I'd like to potentially provide with a fix for this issue.
Since I don't know the code structure at all, I just wanted to double check with you if I'm going on the good direction:

- I understand your feature request is to add an Error-level flag in the Code Inspection everytime the parser finds a New on a class which is annotated as `@Interface`. So basically on Code Inspection run, we parse all `@Interface`s, we check if somewhere there is a `Set xxx = New Interface`, and if so we push an error-level item in the Code Inspection. Am I getting it right? If so, under which category (code quality issue, possible bugs...)?

- From the code structure, under project `RubberDuck.CodeAnalysis.Inspections`, I see that the class `Inspector.cs` runs in parallel a set of inspections enumerated under `IEnumerable<IInspection> inspectionsToRun`. I guess I should add a new class under `Concrete` with this new inspection type and implement the runner in it following the same structure of the other inspections, am I right?
 
- I am debugging the `OptionExplicitInspection` class on a module where I didn't declare `Option Explicit` in order to see what's doing exactly to catch me. Do you feel it's a good starting point for a newbie on this project, or do you suggest me to set my breakpoint somewhere else?

Thanks in advance!
Matteo

**Comment by retailcoder** [comment]
@miannizzotto 

> So, I've cloned the RubberDuck repository

You'll want to have a *fork* of the repository on your GitHub account, and clone the fork - that'll make it easier to integrate your contributions with *pull requests*.

The metadata that all inspections need to do their thing is actually collected *before* the inspections start running - I'll get back to you later tonight with more details, but basically yes, by simply making a new class that inherits one of the base inspection types, the inspector knows to run it - parse tree inspections "walk" parse tree nodes and issue results accordingly; that's mostly useful for things that can be inferred directly from the parse trees, e.g. a redundant `ByRef` modifier, or a missing `Option Explicit` statement. Declaration inspections work off the processed code instead; in Rubberduck pretty much everything is a `Declaration` (think project, module, procedure, variable, etc.), and the resolver traverses the parse trees and does all the hard work of resolving *identifier references* to a specific declaration.

So there are several ways you could go about this - I think the easiest would be to query the declarations to get all user-code interfaces (the `DeclarationFinder` should have methods for that), then iterate their references, and then when the `Context` of an identifier reference is a child of a `SetStmtContext`, you know you've got a new inspection result if the RHS expression has a `NEW` token.

So, I'd suggest looking at a simple inspection that directly inherits `InspectionBase`, like [ConstantNotUsed](https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.CodeAnalysis/Inspections/Concrete/ConstantNotUsedInspection.cs).

The ANTLR grammar is very useful for navigating and casting parse tree nodes; parser rules all turn into `ParserRuleContext`-derived classes, e.g. `setStmt` becomes a `SetStmtContext`:
https://github.com/rubberduck-vba/Rubberduck/blob/71b9554c9d65a85022031818fa5b93449b687139/Rubberduck.Parsing/Grammar/VBAParser.g4#L529

If the `SetStmt` context's `expression()` method yields a `NewExprContext` (inherited from `ExpressionContext`), you know you're looking at code that's trying to new up an interface.

https://github.com/rubberduck-vba/Rubberduck/blob/71b9554c9d65a85022031818fa5b93449b687139/Rubberduck.Parsing/Grammar/VBAParser.g4#L654

Gotta run, sorry if this reply feels rushed - as I said I'll come back in a few hours with more details - in the meantime feel free to join our [dev chat](https://chat.stackexchange.com/rooms/14929/vba-rubberducking) (you have 20+ rep on Stack Exchange? If you don't we can get a moderator to grant you write access anyway), everyone is always thrilled to help new contributors on every/any aspect of contributing!

**Comment by MDoerner** [comment]
Actually, when iterating the references, you should not limit yourself to the descendants of `SetStmtContext`s; you can new up things in a lot more places. Instead you can simply check whether the reference context has an ancestor that is a `VbaParser.NexExprContext`. (There is the extension method `GetAncestor<T>()` yielding the first ancestor of type `T` or `null` if there is none.)

**Comment by bclothier** [comment]
Linking #264

Also, as a possible bonus, we can consider separately flagging the case where an interface has an attribute indicating that it is creatable outside its project. That would only be applicable to VB6 projects but would at least avoid shipping a VB6 DLL that gives out false promises about its external contract.
